index.js

```diff
@@ -1,22 +1,60 @@
 'use strict'
 
+const { ServerResponse } = require('http')
 const fp = require('fastify-plugin')
 const websocket = require('websocket-stream')
+const findMyWay = require('find-my-way')
+
+const kWs = Symbol('ws')
 
 function fastifyWebsocket (fastify, opts, next) {
+  if (opts.handle && typeof opts.handle !== 'function') {
+    return next(new Error('invalid handle function'))
+  }
   const handle = opts.handle
+    ? (req, res) => opts.handle(req[kWs], req)
+    : (req, res) => { req[kWs].socket.close() }
+
   const options = Object.assign({ server: fastify.server }, opts.options)
 
-  if (typeof handle !== 'function') {
-    return next(new Error('invalid handle function'))
-  }
+  const router = findMyWay({
+    ignoreTrailingSlash: true,
+    defaultRoute: handle
+  })
 
-  const wss = websocket.createServer(options, handle)
+  const wss = websocket.createServer(options, handleRouting)
 
   fastify.decorate('websocketServer', wss)
 
+  fastify.addHook('onRoute', routeOptions => {
+    if (routeOptions.method === 'GET') {
+      if (routeOptions.websocket) {
+        if (typeof routeOptions.handler !== 'function') {
+          throw new Error('invalid wsHandler function')
+        }
+        const wsHandler = routeOptions.handler
+        router.on('GET', routeOptions.path, (req, _) => wsHandler(req[kWs], req))
+
+        routeOptions.handler = function (request, reply) {
+          reply.code(404).send()
+        }
+      } else if (routeOptions.wsHandler) {
+        if (typeof routeOptions.wsHandler !== 'function') {
+          throw new Error('invalid wsHandler function')
+        }
+        router.on('GET', routeOptions.path, (req, _) => routeOptions.wsHandler(req[kWs], req))
```

Comment:
Can you dry the two call to the router in a helper function?

---

Also, you'll need to handle the case of `wsHandler`  returning a promise. If that errors, we should destroy the websocket connection. Otherwise we will likely leak memory.